array_api_tests/test_type_promotion.py

```diff
@@ -1,738 +1,242 @@
-    binary_op = binary_operators[binary_op_name]
-
-    (type1, type2), res_type = dtypes
-    dtype1 = dtype_mapping[type1]
-    dtype2 = dtype_mapping[type2]
-    res_dtype = dtype_mapping[res_type]
-    fillvalue1 = fillvalues.draw(scalars(just(dtype1)))
-    if binary_op_name in ['>>', '<<']:
-        fillvalue2 = fillvalues.draw(scalars(just(dtype2)).filter(lambda x: x > 0))
+        arrays = []
+        shapes = data.draw(
+            hh.mutually_broadcastable_shapes(len(in_dtypes)), label='shapes'
+        )
+        for i, (dtype, shape) in enumerate(zip(in_dtypes, shapes), 1):
+            x = data.draw(
+                xps.arrays(dtype=dtype, shape=shape).filter(x_filter),
+                label=f'x{i}',
+            )
+            arrays.append(x)
+        out = func(*arrays)
+    assert out.dtype == out_dtype, f'{out.dtype=!s}, but should be {out_dtype}'
+
+
+def gen_op_params() -> Iterator[Tuple[str, Tuple[DT, ...], DT, Callable]]:
+    op_to_symbol = {**dh.unary_op_to_symbol, **dh.binary_op_to_symbol}
+    for op, symbol in op_to_symbol.items():
+        if op == '__matmul__':
+            continue
+        in_category = dh.op_in_categories[op]
+        out_category = dh.op_out_categories[op]
+        valid_in_dtypes = dh.category_to_dtypes[in_category]
+        ndtypes = nargs(op)
+        if ndtypes == 1:
+            for in_dtype in valid_in_dtypes:
+                out_dtype = in_dtype if out_category == 'promoted' else xp.bool
+                yield pytest.param(
+                    f'{symbol}x',
+                    (in_dtype,),
+                    out_dtype,
+                    filters[op],
+                    id=f'{op}({in_dtype}) -> {out_dtype}',
+                )
+        else:
+            for (in_dtype1, in_dtype2), promoted_dtype in dh.promotion_table.items():
+                if in_dtype1 in valid_in_dtypes and in_dtype2 in valid_in_dtypes:
+                    out_dtype = (
+                        promoted_dtype if out_category == 'promoted' else xp.bool
+                    )
+                    yield pytest.param(
+                        f'x1 {symbol} x2',
+                        (in_dtype1, in_dtype2),
+                        out_dtype,
+                        filters[op],
+                        id=f'{op}({in_dtype1}, {in_dtype2}) -> {out_dtype}',
+                    )
+    # We generate params for abs seperately as it does not have an associated symbol
+    for in_dtype in dh.category_to_dtypes[dh.op_in_categories['__abs__']]:
+        yield pytest.param(
+            'abs(x)',
+            (in_dtype,),
+            in_dtype,
+            filters['__abs__'],
+            id=f'__abs__({in_dtype}) -> {in_dtype}',
+        )
+
+
+@pytest.mark.parametrize('expr, in_dtypes, out_dtype, x_filter', gen_op_params())
+@given(data=st.data())
+def test_operator_returns_array_with_correct_dtype(
+    expr, in_dtypes, out_dtype, x_filter, data
+):
+    if len(in_dtypes) == 1:
+        x = data.draw(
```

Comment:
Is it possible to avoid `draw` here, or can strategies not access parameterization args?

Reply:
Yeah I'm pretty sure we have to draw dynamically, and if not then it'd probably be too much of a hack anyway. Also the labels should make it a bit clearer what arrays exactly cause problems when Hypothesis finds failures.